### ISSUE TITLE
Avoid unnecessary roundtrips

### ISSUE THREAD

**Issue by pekeler** [arangodb/arangojs#52]
I've been wondering for a while why collection() is an asynchronous method. It seems that it actually makes a roundtrip to the DB. But why?

```JavaScript
db.collection('MyThings')
  .then(collection => collection.count())
```
So to get the number of MyThings it makes two roundtrips even though, theoretically, it should be just one?  Or if I want to insert one document each into 10 different collections, I'd end up with 20 roundtrips?

**Comment by pluma** [comment]
It makes the roundtrip because it needs to verify the collection actually exists and to determine whether it is a document or edge collection (which determines what API it exposes).

If you need to insert one document into 10 different collections regularly, you should probably either use AQL or write a Foxx app -- even without the overhead of fetching each collection the number of requests seems impractical.

**Comment by pekeler** [comment]
I can't imagine a use case where I'd want to check for existence of collection before doing something with it, or where I didn't know the type of the collection. Unless I want to build a DB-management tool. (Is Aardvark using this adapter? That would explain why it's working this way.)

I'd say that the typical use case for this adapter is to access ArangoDB remotely from a client which has certain expectations about the DB. And with this use case, I want to minimize the number of roundtrips because they're typically the bottleneck.

Getting an exception when trying to access a collection which doesn't exist anymore is IMHO perfectly acceptable. As is using different methods to access edge collections.

**Comment by pluma** [comment]
cc @fceller 

I can see your point, but I'm not entirely sure how to best approach this.

Following your argument the changes probably should be:

* db.collection -> sync, always returns DocumentCollection, remove "autoCreate" flag
* db.edgeCollection -> sync, always returns EdgeCollection, remove "autoCreate" flag
* db.database -> sync, remove "autoCreate" flag
* db.graph -> sync, remove "autoCreate" flag
* collection.properties, etc -> remove the side-effects that modify the collection object

However this would be a breaking change (semver-major). We should probably use the opportunity to decide whether there are any other backwards-incompatible changes we want to make.

**Comment by pekeler** [comment]
Yea, I also couldn't imagine ever needing the autoCreate flags.

Perhaps it would make sense to have two adapters. This one for people that need the flexibility. And a new, fast one :)

Or maybe a survey to find out if there actually is anybody needing the flexibility of this current version.

**Comment by pluma** [comment]
I think the `autoCreate` logic was inherited from the Ruby driver. As this is explicitly intended to be a low-level driver (something you can build higher-level drivers and ODMs on), I wouldn't really miss it. It might just make some tutorials a bit wordier.

**Comment by frankmayer** [comment]
Why not manually declare the collection specifics, if we know them already? 
Based on those specifics, we would get a collection instance to work with. 
This would probably not require big changes, which would result in aforementioned BC breaks.

By passing the name and whatever else is necessary (like, for example, the type) it should be possible to manually construct the instance, bypassing the round-trip.

It could look somewhat like this:
```
db.constructCollectionInstance('my-collection',{'type: 2 |+ whatever else is needed|}, function (err, collection) {
    if (err) 
        // something went wrong... 
        console.error(err);
        return;
    }
    // collection exists
});)
```

What do you think?

**Comment by frankmayer** [comment]
Following up on my previous comment, I found myself some time to do a quick and dirty implementation of that proposal. Since I am not sure if you guys want to follow that path, I am not making a pull request out of that. However, here is what I did to make it work.

add this in the ```database.js``` file inside the ```extend(Database.prototype, {``` declaration.
```
  constructCollectionInstance : function constructCollectionInstance (collectionName, specifics, cb) {
    specifics.name = collectionName;

    var _connection$promisify3 = this._connection.promisify(cb);
    var promise = _connection$promisify3.promise;
    var callback = _connection$promisify3.callback;
    callback(null, _createCollection(this._connection, specifics));
    return promise;
  },

```

Now you can create a collection instance by just calling the proposed method (here using the promise):
```
    var collection = yield db.constructCollectionInstance('my-collection',{type: 2})
        .then(function (collection) {
            return collection;
        });
```

Performance-wise, it might be better, though, to just persist and reuse the collection. I guess, in the end, it depends on the use case.

**Comment by pluma** [comment]
@frankmayer There are already `createCollection` (and `createEdgeCollection`) methods. I'm not sure what your proposal would improve on that.

The entire `autoCreate` logic was lifted from the Ruby driver and is probably out of scope for this project as the design goal is to have a high-performance low-level driver that only provides abstractions where they are necessary.

**Comment by frankmayer** [comment]
@pluma, my proposal has nothing to do with creating collection (on the server). As I wrote, it provides the functionality to get rid of the unnecessary callback in order to get the collection instance, needed to work with documents, for example.

As for the autocreate logic & abstractions you mentioned: 
I believe, the low level driver should encapsulate the http API of the server. This means, that it would also bea able to pass on the autoCreate flag as an option, as it it provided by the API. Such option-collections should probably be objects instead of specific parameters to functions.

**Comment by pluma** [comment]
I'm making some modifications to the driver that removes the autoCreate functionality completely and has the collection/edgeCollection/graph methods return instantly rather than making separate requests.

You can track the progress in the 4.0 branch: https://github.com/arangodb/arangojs/tree/4.0

**Comment by pekeler** [comment]
@pluma : sweet :)

**Comment by pluma** [comment]
Fixed in upcoming 4.0 release.

**Comment by kevinprotoss** [comment]
Hi guys,

Sorry, I have a different opinion about this topic.

Why `autoCreate` is not required? For our use case, the use or creation of collection could be dynamic. For some kinds of database as a service, `autoCreate` for database level is also required. I do not understand about this change. I believe ArangoDB still support this flag, doesn't it? If yes, why the driver doesn't?

If using v4, how to make sure the database / collection has been created? Must I manually check and create? Will this solution get a better performance than before?

Bg, 
Kevin

**Comment by kevinprotoss** [comment]
To avoid unnecessary roundtrips, the driver just makes one roundtrip to DB iff autoCreate flag is set to true. It's an optional flag, isn't it? Why completely remove that?

**Comment by pluma** [comment]
The old autoCreate behaviour can easily be implemented like this (1207 indicates "duplicate collection name"):

```js
let collection = db.collection('some-collection');

collection.create()
.catch(err => {
  if (err.code !== 1207) return Promise.reject(err);
})
.then(() => {
  // collection definitely exists now
});

Or if you need multiple collections:

let collections = [
  db.collection('one'),
  db.collection('two')
];

Promise.all(collections.map(
  collection => collection.create()
  .catch(err => {
    if (err.code !== 1207) return Promise.reject(err);
  })
))
.then(() => {
  // collection definitely exists now
});
```

In most usage scenarios you either expect to use an existing collection (where the collection missing would indicate some kind of problem) or you want to create new collections (where the collection already existing would indicate some kind of problem).

Making the `collection`/`useDatabase`/`graph`/etc methods synchronous drastically simplifies common scenarios: creating Collection/Graph/etc objects no longer requires making sure they actually exist, they just represent the respective APIs (just like `new Database` doesn't create a database on the server, it just creates a new wrapper for the API of that database).

Lastly, `autoCreate` is actually better implemented as "ignore duplicate" (as in my example). The 3.x implementation performed three requests:

1. Fetch the metadata, which fails because it doesn't exist yet
2. Create it with the default settings (and ignore duplicate to avoid concurrency issues)
3. Fetch the metadata again because the response for create is insufficient

In other words: if you don't actually need the metadata (the name is all you need to use the API) you would end up with one unnecessary request in the best case (first request succeeds) and two in the worst case (first request fails).

We could have added a "make sure this exists" method as a substitute for the old `autoCreate` option but that would have forced us to make decisions that have nothing to do with the driver, like how to deal with the lack of a useful server response if creation fails, or whether it should actually be an upsert (i.e. if creation fails, try to modify the collection's/graph's/database's properties to match the options).

**Comment by kevinprotoss** [comment]
@pluma Thanks for your answer.

I understand it's easy to implement this logic by myself. ArangoDB provides a HTTP based API. I could do anything as I like of course, but I still think the driver could have such a feature if possible.

**Comment by pluma** [comment]
True. You could do all the HTTP calls yourself, but we feel that a low-level driver such as this one provides enough value simply by abstracting away most of the actual HTTP interactions (e.g. which routes map to what actions) and providing more optimized performance.

If you feel that having such an API call for collections, graphs or databases would be valuable to you and other users I suggest you propose its addition to the HTTP API by opening an issue on the ArangoDB repository itself. If the feature is useful enough to be implemented in a driver, it's likely useful enough to be part of the proper HTTP API (which would also avoid the additional HTTP calls).
